CORE-8415 - Upgrade Javalin to version 6.2.0

filipesoliveira opened the pull request:
This PR depends on:
-  https://github.com/javalin/javalin/pull/2272. (PR has been merged into master. Version 6.2.0 of Javalin has been released.)
- https://github.com/corda/corda-runtime-os/pull/6248#event-13366868114  (PR has been merged into master).

TODO list:

- [x] 1) Once the PR is green, the e2e tests must be run. (Done.)
- [x] 2) Once the e2e tests are green, the Q.A. regression tests must be run. (Done.)
- [x] 3) Once all tests are green, remove the temporary repo that was added for testing purposes. (Done.)
- [x] 4) Update the Javalin version to point to an official release instead of a snapshot. (Done. Javalin 6.2.0.)

filipesoliveira committed:
CORE-8415 - Added a OSGi complaint Javalin 6.1.7 binaries. This is only for testing purpose only and the binaries will be removed before the changes are merged.

filipesoliveira committed:
Updated the code to be compatible with Javalin 6.1.7.

filipesoliveira commented on test-repo/io/javalin/javalin-osgi/maven-metadata-local.xml:
This module must be removed before merging the PR.

filipesoliveira committed:
Reverted changes.

filipesoliveira committed:
Updated the path to the `test-repo`

filipesoliveira committed:
Added `javalin-parent` to the `test-repo`.

filipesoliveira committed:
Added all javalin related artifacts to the `test-repo`.

filipesoliveira committed:
Removed unnecessary jars.

filipesoliveira committed:
Removed .idea folder.

filipesoliveira committed:
Removed gradle related file that are not necessary.

vkolomeyko commented:
I can see compilation error as follows:
```
:tools:corda-runtime-gradle-plugin:compileIntegrationTestKotlin FAILED	
e: file:///home/jenkins/agent/workspace/Corda5_corda-runtime-os_PR-6211/tools/corda-runtime-gradle-plugin/src/integrationTest/kotlin/net/corda/gradle/plugin/FunctionalBaseTest.kt:15:44 No value passed for parameter 'p0'
```
Let's keep PR as Draft till we get a green build.

vkolomeyko commented on gradle/libs.versions.toml:
```diff
@@ -9,11 +9,12 @@ caffeineVersion = "3.1.8"
 # Once Javalin version is upgraded to the latest, this override may be removed.
-jettyVersion = "9.4.54.v20240208"
+jettyVersion = "11.0.16"
+jettyWebsocketServerVersion = "9.4.51.v20230217"
```
```suggestion
jettyWebsocketClientVersion = "9.4.54.v20240208"
```
https://mvnrepository.com/artifact/org.eclipse.jetty.websocket/websocket-client/9.4.54.v20240208

filipesoliveira committed:
Fixed compilation issues.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Fixed unit tests.

filipesoliveira committed:
Reverted changem

filipesoliveira committed:
Reverted change.

filipesoliveira committed:
Reverted change regarding a test

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Update

filipesoliveira committed:
Update

vkolomeyko commented on libs/rest/rest-server-impl/src/main/kotlin/net/corda/rest/server/impl/internal/RestServerInternal.kt:
```diff
@@ -46,8 +48,7 @@ import org.osgi.framework.FrameworkUtil
 import java.nio.file.Path
-import java.util.LinkedList
-import javax.servlet.MultipartConfigElement
+import java.util.*
```
We tend not to use wildcard imports. An option in IntelliJ probably needs tweaking for it to stop doing that during Optimize Imports operation.

vkolomeyko commented on libs/rest/rest-server-impl/src/test/kotlin/net/corda/rest/server/impl/RestServerTest.kt:
```diff
@@ -42,7 +42,7 @@ class RestServerTest {
         doReturn(false).whenever(configProvider).isDevModeEnabled()
         Assertions.assertThrows(
-            IllegalArgumentException::class.java,
+            Exception::class.java,
```
Out of curiosity, how come that exception type got changed on this instance?

filipesoliveira replied:
This is a strange one. I am not sure how this test ever passed because I can see the `IllegalArgumentException` being wrapped with an Exception instance.

I will come back to this change later on.

vkolomeyko commented on libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:
```diff
@@ -42,11 +42,8 @@ class JavalinServer(
+                config.http.maxRequestSize = maxRequestSize
 
-                if (log.isDebugEnabled) {
-                    config.enableDevLogging()
```
Is there no equivalent call in the new Javalin to enable additional logging of its internals?

filipesoliveira replied:
Yes, there is an equivalent and the code will be added back.

filipesoliveira committed:
Removed import with wildcard.

filipesoliveira committed:
Added the `enableDevLogging` call back to the Javalin server.

filipesoliveira committed:
Reverted change to fix failing tests.

filipesoliveira committed:
Reverted a change.

filipesoliveira committed:
Updated the method `fromJsonString`. If the string "null" is passed in as a parameter, `null` is returned immediately.

filipesoliveira committed:
Resolved issue reported by detekt.

filipesoliveira committed:
Removed extra white line.

filipesoliveira committed:
Reverted change.

filipesoliveira committed:
Fixed issue reported by detekt

filipesoliveira committed:
Disable a test.

filipesoliveira committed:
Updated the SSL configuration and fixed an unstable test.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
The class `DuplexChannel` now sends messages synchronously.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Commented out a couple of asserts. They will be added back later.

filipesoliveira committed:
Disabled a test related to swagger. Will come back later.

filipesoliveira committed:
Disabled a test related to swagger. Will come back later.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira commented on libs/rest/rest-server-impl/build.gradle:
```diff
@@ -29,16 +29,11 @@ dependencies {
     implementation project(":libs:rest:rest-server")
     implementation "net.corda:corda-crypto"
     implementation libs.javalin
-    constraints {
```
Jetty has been upgraded to version 11.x.

filipesoliveira commented on libs/rest/rest-server-impl/src/main/kotlin/net/corda/rest/server/impl/websocket/ServerDuplexChannel.kt:
```diff
@@ -18,12 +17,12 @@ internal class ServerDuplexChannel(
-    override fun send(message: String): Future<Void> {
-        return ctx.send(message)
+    override fun send(message: String) {
+        ctx.send(message)
```
The send method is now synchronous after upgrading Javelin from version 4 to 6.

filipesoliveira commented on libs/rest/rest-test-common/src/main/kotlin/net/corda/rest/test/TestHealthCheckAPIImpl.kt:
```diff
@@ -118,11 +118,9 @@ class TestHealthCheckAPIImpl : TestHealthCheckAPI, PluggableRestResource<TestHea
                 {
                     log.debug { "Sending: $counter" }
-                    val future = channel.send("${counter++}")
+                    channel.send("${counter++}")
```
Before the send method was asynchronous. Now it is a synchronous call.

filipesoliveira committed:
Reverted a change.

filipesoliveira committed:
Upgraded `networkntJsonSchemaVersion` to the most recent version that does not have breaking changes.

filipesoliveira commented on gradle.properties:
```diff
@@ -59,7 +59,7 @@ kotlinCoroutinesVersion=1.6.4
 beanutilsVersion=1.9.4
 # com.networknt:json-schema-validator cannot be upgraded beyond 1.0.79 because it requires an OSGi bundle containing SLF4j with a version in the range [2.0,3), whereas Corda currently provides 1.7.36.
-networkntJsonSchemaVersion = 1.0.79
+networkntJsonSchemaVersion = 1.2.0
```
`NetworkntJsonSchema` was updated to the most recent version that does not have breaking changes. Updating the version was necessary to fix build issues after upgrading to Jackson 6.

filipesoliveira replied:
This change is not required.

filipesoliveira committed:
Reverted to the original version for networkntJsonSchemaVersion.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira commented on build.gradle:
```diff
@@ -151,6 +151,13 @@ allprojects {
+        // CORE-8415 - This repo is for testing purpose only and must be removed before the
+        // code is merged into master
+        maven {
+            url = "file://${rootDir}/test-repo"
```
This must be removed.

filipesoliveira commented on gradle/libs.versions.toml:
```diff
@@ -9,11 +9,11 @@ caffeineVersion = "3.1.8"
 hikariCpVersion = "5.1.0"
 jacksonVersion = "2.17.1"
-javalinVersion = "4.6.8"
+javalinVersion = "6.1.7-SNAPSHOT"
```
This must be updated once a version of Javalin that supports OSGi is released.

filipesoliveira replied:
Javalin 6.2.0 has been released.

filipesoliveira committed:
Fixed failing test due to the charset not being set.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Replace `.name` by `.mimeType`.

filipesoliveira committed:
The protocol is required in the origin for the headers in the response to be set correctly.

filipesoliveira committed:
Fixed issue reported by detekt.

filipesoliveira committed:
Updated the `networkntJsonSchemaVersion` to fix issue with the versioning of snakeyaml.

filipesoliveira committed:
Revert "Updated the `networkntJsonSchemaVersion` to fix issue with the versioning of snakeyaml."

This reverts commit 90d41cba224ad65928b6b2c1298a48a6fcc27c5e.

filipesoliveira committed:
Added the response encoding to the request to avoid the charset to be set to null when requesting `swagger-ui-bundle.js`.

filipesoliveira committed:
Made the package `org.slf4j.spi` optional in an attempt to resolve a collision.

filipesoliveira committed:
Quick test. Revert this change if the build fails.

filipesoliveira committed:
Revert "Quick test. Revert this change if the build fails."

This reverts commit f2427cee85930058d701d6211db8e5716804c551.

filipesoliveira committed:
CORE-8415: Filter out unwanted SLF4J bundle in tests

Squashed commit of the following:

commit ab116fc4656b71fca584b92a3c6cce07526ebb26
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Tue Jul 2 09:21:47 2024 +0100

    CORE-8415: Fix sandbox setup to use test bundle of SLF4J

commit 38fa2c387942c121217619bdb3cb88aae0527c27
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Tue Jul 2 07:56:25 2024 +0100

    CORE-8415: Include `net.corda.slf4jv1` bundle as a requirement for OSGi tests

commit 90ac86dcfabd134ee104736acf182974c27f1700
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Mon Jul 1 21:51:47 2024 +0100

    CORE-8415: More fixes to make `resolve` task pass

commit 21b8e229d7b7e889285c4c0aaa14a87b213bb9f2
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Mon Jul 1 20:55:13 2024 +0100

    CORE-8415: More fixes to integrationTestRuntime configuration

commit 7375f23bc449f4e4694b39c8417a4ee236d3e260
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Mon Jul 1 18:40:54 2024 +0100

    CORE-8415: More fixes to integrationTestRuntime configuration

commit e845e540dcd34cda86fa956591fbd20b75085b1c
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Mon Jul 1 18:33:01 2024 +0100

    CORE-8415: More logging to narrow down the problem

commit bef8ff3e9aebe43c2550a912c4c5a9eb716fec87
Author: Viktor Kolomeyko <Viktor.Kolomeyko@r3.com>
Date:   Mon Jul 1 17:51:47 2024 +0100

    CORE-8415: Try to filter out unwanted dependency

    For OSGi tests only

filipesoliveira commented:
build e2e

filipesoliveira committed:
Removed comments.

filipesoliveira commented:
build e2e

filipesoliveira commented:
build e2e

filipesoliveira commented:
The e2e tests were green after fixing the issue affecting the two tests. 

Link to the green build - https://ci02.dev.r3.com/blue/organizations/jenkins/Corda5%2Fcorda5-e2e-tests/detail/PR-665/6/pipeline

filipesoliveira committed:
Reverted a change made to a test.

filipesoliveira committed:
Simplified the code responsible for the basic authentication.

filipesoliveira committed:
Removed unnecessary import.

filipesoliveira committed:
Removed unncessary imports.

filipesoliveira committed:
Fixed unit test related to the basic auth.

filipesoliveira committed:
Fixed unit tests related to the basic auth.

filipesoliveira committed:
Fixed issues reported by detekt.

filipesoliveira committed:
Updated Jetty to the lastest version of the major version 11.

filipesoliveira commented:
Q.A. regression tests are green.

Link to the green build - https://ci05.dev.r3.com/job/QA/job/CRAFT5/job/Craft_Tests/job/C5/job/qa-regpack-tests/job/PR-1153/

filipesoliveira commented:
The changes to enable OSGi in Javalin 6 have been merged. The version 6.1.7 of Javalin 6 will support OSGi.

References:
1. PR that enables OSGi in Javalin 6 : https://github.com/javalin/javalin/pull/2272
2. Releases of Javalin 6: https://github.com/javalin/javalin/releases

filipesoliveira committed:
Fixed build issue after updating the branch with the latest changes from the branch `release/os/5.3`

filipesoliveira commented:
A new version of Javalin has been released.

Reference:
- Javalin 6.2.0 release: https://github.com/javalin/javalin/releases

filipesoliveira committed:
Updated the version of Javalin to 6.2.0.

filipesoliveira committed:
Removed the `test-repo` used for testing javalin-6.1.7-SNAPSHOT.

filipesoliveira commented:
build e2e

filipesoliveira commented:
build e2e

tlawson3 reviewed (dismissed):
LGTM - Tooling files

vkolomeyko commented on libs/rest/rest-server-impl/src/main/kotlin/net/corda/rest/server/impl/internal/RestServerInternal.kt:
```diff
@@ -91,36 +89,42 @@ internal class RestServerInternal(
             addWsRoutes()
             // In order for multipart content to be stored onto disk, we need to override some properties
             // which are set by default by Javalin such that entire content is read into memory
+            @Suppress("DEPRECATION")
```
I thought I made a change to get rid of thsi deprecation:
https://github.com/corda/corda-runtime-os/pull/6285/files

filipesoliveira replied:
I will add that change back. Not sure what happen.

filipesoliveira committed:
Added back a change purposed by Viktor that had been lost.

filipesoliveira commented:
build e2e

vkolomeyko reviewed (approved):
LGTM

tlawson3 reviewed (approved):
LGTM - tooling
